File: Greedy/Java/Lemonade Change.java
// Problem link: https://leetcode.com/problems/lemonade-change/description
// Time: O(n)
// Space: O(1)
class Solution {
    public boolean lemonadeChange(int[] bills) {
        int ten = 0, five = 0;

        for(int dollar : bills){
            if(dollar == 5){
                five++;
            }
            else if(dollar == 10){
                if(five == 0){
                    return false;
                }
                five--;
                ten++;
            }
            else{
                if(ten > 0 && five > 0){
                    ten--;
                    five--;
                }
                else if(five > 2){
                    five -= 3;
                }
                else{
                    return false;
                }
            }
        }
        return true;
    }
}

File: Greedy/Python/Lemonade Change.py
# Problem link: https://leetcode.com/problems/lemonade-change/description/
# Time: O(n)
# Space: O(1)
class Solution:
    def lemonadeChange(self, bills: List[int]) -> bool:
        store = {10 : 0, 5 : 0}

        for i in bills:
            if i == 5:
                store[5] += 1
            elif i == 10:
                if store[5] == 0:
                    return False
                store[5], store[10] = store[5] - 1, store[10] + 1
            else:
                if store[10] > 0 and store[5] > 0:
                    store[5], store[10] = store[5] - 1, store[10] - 1
                elif store[5] > 2:
                    store[5] -= 3
                else:
                    return False
        
        return True
